pipeline/backend/dummy/dummy.go

```diff
@@ -0,0 +1,234 @@
+	return nil
+}
+
+func (e *dummy) StartStep(_ context.Context, step *backend.Step, taskUUID string) error {
+	log.Trace().Str("taskUUID", taskUUID).Msgf("start step %s", step.Name)
+
+	// internal state checks
+	_, exist := e.kv.Load("task_" + taskUUID)
+	if !exist {
+		return fmt.Errorf("expect env of workflow %s to exist but found none to destroy", taskUUID)
+	}
+	stepState, stepExist := e.kv.Load(fmt.Sprintf("task_%s_step_%s", taskUUID, step.UUID))
+	if stepExist {
+		// Detect issues like https://github.com/woodpecker-ci/woodpecker/issues/3494
+		return fmt.Errorf("StartStep detected already started step '%s' (%s) in state: %s", step.Name, step.UUID, stepState)
+	}
+
+	if step.Name == StepStartFail {
+		return fmt.Errorf("expected fail to start step")
+	}
+
+	expectStepType, testStepType := step.Environment[EnvKeyStepType]
+	if testStepType && string(step.Type) != expectStepType {
+		return fmt.Errorf("expected step type '%s' but got '%s'", expectStepType, step.Type)
+	}
+
+	e.kv.Store(fmt.Sprintf("task_%s_step_%s", taskUUID, step.UUID), stepStateStarted)
+	return nil
+}
+
+func (e *dummy) WaitStep(ctx context.Context, step *backend.Step, taskUUID string) (*backend.State, error) {
+	log.Trace().Str("taskUUID", taskUUID).Msgf("wait for step %s", step.Name)
+
+	_, exist := e.kv.Load("task_" + taskUUID)
+	if !exist {
+		err := fmt.Errorf("expect env of workflow %s to exist but found none to destroy", taskUUID)
+		return &backend.State{Error: err}, err
+	}
+
+	// check state
+	stepState, stepExist := e.kv.Load(fmt.Sprintf("task_%s_step_%s", taskUUID, step.UUID))
+	if !stepExist {
+		err := fmt.Errorf("WaitStep expect step '%s' (%s) to be created but found none", step.Name, step.UUID)
+		return &backend.State{Error: err}, err
+	}
+	if stepState != stepStateStarted {
+		err := fmt.Errorf("WaitStep expect step '%s' (%s) to be '%s' but it is: %s", step.Name, step.UUID, stepStateStarted, stepState)
+		return &backend.State{Error: err}, err
+	}
+
+	// extend wait time logic
+	if sleep, sleepExist := step.Environment[EnvKeyStepSleep]; sleepExist {
+		toSleep, err := time.ParseDuration(sleep)
+		if err != nil {
+			err = fmt.Errorf("WaitStep fail to parse sleep duration: %w", err)
+			return &backend.State{Error: err}, err
+		}
+		time.Sleep(toSleep)
+	} else {
+		if step.Type == backend.StepTypeService {
+			select {
+			case <-time.NewTimer(testServiceTimeout).C:
+				err := fmt.Errorf("WaitStep fail due to timeout of service after 1 second")
+				return &backend.State{Error: err}, err
+			case <-ctx.Done():
+				// context for service closed ... we can move forward
+			}
+		} else {
+			time.Sleep(time.Nanosecond)
+		}
+	}
+
+	e.kv.Store(fmt.Sprintf("task_%s_step_%s", taskUUID, step.UUID), stepStateDone)
+
+	if step.Name == StepExecError {
+		return &backend.State{
+			ExitCode:  1,
+			Exited:    true,
+			OOMKilled: false,
+		}, nil
+	}
+
+	return &backend.State{
+		ExitCode:  0,
+		Exited:    true,
+		OOMKilled: false,
+	}, nil
```

Comment:
Same thing as before that only one step can be set to fail as it otherwise can't be detected in logs  etc.

```suggestion
  oomKilled := step.Environment[EnvKeyOOMKilled] == "true"
	exitCode := 0
	if exitCodeStr, exitCodeExist := step.Environment[EnvKeyExitCode]; exitCodeExist {
		exitCode, _ = strconv.Atoi(exitCodeStr)
	}

	return &backend.State{
		ExitCode:  exitCode,
		Exited:    true,
		OOMKilled: oomKilled,
	}, nil
```